replicator/main.go

```diff
@@ -0,0 +1,847 @@
+	tableName   string
+	consistency gocql.Consistency
+
+	pkColumns    []string
+	ckColumns    []string
+	otherColumns []string
+	columnTypes  map[string]TypeInfo
+	allColumns   []string
+
+	rowDeleteQueryStr       string
+	partitionDeleteQueryStr string
+	rangeDeleteQueryStrs    []string
+
+	localCount int64
+	totalCount *int64
+}
+
+type updateQuerySet struct {
+	add    string
+	remove string
+}
+
+type udtInfo struct {
+	setterQuery string
+	fields      []string
+}
+
+func NewDeltaReplicator(session *gocql.Session, kmeta *gocql.KeyspaceMetadata, meta *gocql.TableMetadata, consistency gocql.Consistency, count *int64) (*DeltaReplicator, error) {
+	var (
+		pkColumns    []string
+		ckColumns    []string
+		otherColumns []string
+	)
+
+	for _, name := range meta.OrderedColumns {
+		colDesc := meta.Columns[name]
+		switch colDesc.Kind {
+		case gocql.ColumnPartitionKey:
+			pkColumns = append(pkColumns, name)
+		case gocql.ColumnClusteringKey:
+			ckColumns = append(ckColumns, name)
+		default:
+			otherColumns = append(otherColumns, name)
+		}
+	}
+
+	columnTypes := make(map[string]TypeInfo, len(meta.Columns))
+	for colName, colMeta := range meta.Columns {
+		info := parseType(colMeta.Type)
+		columnTypes[colName] = info
+	}
+
+	dr := &DeltaReplicator{
+		session:     session,
+		tableName:   meta.Keyspace + "." + meta.Name,
+		consistency: consistency,
+
+		pkColumns:    pkColumns,
+		ckColumns:    ckColumns,
+		otherColumns: otherColumns,
+		columnTypes:  columnTypes,
+		allColumns:   append(append(append([]string{}, otherColumns...), pkColumns...), ckColumns...),
+
+		totalCount: count,
+	}
+
+	dr.computeRowDeleteQuery()
+	dr.computePartitionDeleteQuery()
+
+	return dr, nil
+}
+
+func (r *DeltaReplicator) computeRowDeleteQuery() {
+	keyColumns := append(append([]string{}, r.pkColumns...), r.ckColumns...)
+
+	r.rowDeleteQueryStr = fmt.Sprintf(
+		"DELETE FROM %s WHERE %s",
+		r.tableName,
+		r.makeBindMarkerAssignments(keyColumns, " AND "),
+	)
+}
+
+func (r *DeltaReplicator) computePartitionDeleteQuery() {
+	r.partitionDeleteQueryStr = fmt.Sprintf(
+		"DELETE FROM %s WHERE %s",
+		r.tableName,
+		r.makeBindMarkerAssignments(r.pkColumns, " AND "),
+	)
+}
+
+func (r *DeltaReplicator) Consume(c scylla_cdc.Change) error {
+	timestamp := c.GetCassandraTimestamp()
+	pos := 0
+
+	if showTimestamps {
+		log.Printf("Processing timestamp: %v\n", c.Time)
+	}
+
+	for pos < len(c.Delta) {
+		change := c.Delta[pos]
+		var err error
+		switch change.GetOperation() {
+		case scylla_cdc.Update:
+			err = r.processUpdate(timestamp, change)
+			pos++
+
+		case scylla_cdc.Insert:
+			err = r.processInsert(timestamp, change)
+			pos++
+
+		case scylla_cdc.RowDelete:
+			err = r.processRowDelete(timestamp, change)
+			pos++
+
+		case scylla_cdc.PartitionDelete:
+			err = r.processPartitionDelete(timestamp, change)
+			pos++
+
+		case scylla_cdc.RangeDeleteStartInclusive, scylla_cdc.RangeDeleteStartExclusive:
+			// TODO: Check that we aren't at the end?
```

Comment:
On the other hand, since there are two entries - not necessarily
but we can make a sanity check. I'd even make it stronger: not only check that we're not on the end, but also that the next pos is a range end entry. But I don't know what should be done if the entry is gone. I see two options:
1. finish with error
2. log error but continue, assuming one sided range

I guess finish with error is the safer option since we don't want to "silently" introduce inconsistencies.
On the other hand, someone may want to continue regardless (go with option 2)...

Reply:
Option 2 seems more dangerous, because if range end is missing due to an inconsistency, it will delete more rows than necessary. I think I'll go with option 1 for now.

Reply:
Now, I'm properly checking that range delete start and range delete end rows are always paired, and stop with an error if they are not.